sklearn/utils/_array_api.py

```diff
@@ -545,18 +605,39 @@ def _fill_or_add_to_diagonal(array, value, xp, add_value=True, wrap=False):
         )
 
     value = xp.asarray(value, dtype=array.dtype, device=device(array))
-    end = None
-    # Explicit, fast formula for the common case.  For 2-d arrays, we
-    # accept rectangular ones.
+    if value.ndim not in [0, 1]:
+        raise ValueError(
+            "value needs to be a scalar or a 1d array, "
+            f"got a {value.ndim}d array instead"
+        )
+
     step = array.shape[1] + 1
-    if not wrap:
-        end = array.shape[1] * array.shape[1]
+    # 'end' avoids wrapping in case of array is non-square with n_rows > n_columns
+    end = array.shape[1] * array.shape[1]
+    min_rows_columns = min(array.shape)
 
-    array_flat = xp.reshape(array, (-1,))
-    if add_value:
-        array_flat[:end:step] += value
+    if _is_numpy_namespace(xp):
+        assignment_function(array.flat, slice(None, end, step), value)
```

Comment:
I've realised that this implementation technically differs from our loop implementation, because if `value` was 1D but shorter than the diagonal, `value` would be repeated to fill all entries, as is done in [numpy docs](https://numpy.org/doc/stable/reference/generated/numpy.fill_diagonal.html):

> If array-like, the flattened val is written along the diagonal, repeating if necessary to fill all diagonal entries.

Though looking at all our use cases, `value` should always be the correct size.

With the loop implementation, there is no repeating.

I am tempted to use the loop, even for numpy namespace, just to ensure consistent results...